znnl/data/decision_boundary.py

```diff
@@ -0,0 +1,193 @@
+    reference_values = gradient * data[:, 0] + intercept
+
+    differences = data[:, 1] - reference_values
+
+    differences[differences > 0] = 1
+    differences[differences < 0.0] = 0
+
+    return differences
+
+
+def circle(data: onp.ndarray, radius: float = 0.25):
+    """
+    Create a circular classification problem.
+
+    For simplicity, assume the points inside the circle are
+    class 0 and outside are class 1.
+
+    Parameters
+    ----------
+    data : np.ndarray
+        Data to be converted into classes.
+    radius : float
+        Radius of the circle.
+    """
+    radii = onp.linalg.norm(data - 0.5, axis=1)
+
+    radii[radii < radius] = 0.0
+    radii[radii > radius] = 1.0
+
+    return radii
+
+
+class DecisionBoundaryGenerator(DataGenerator):
+    """
+    A class to generate data for decision boundary problems.
+    """
+
+    def __init__(
+        self,
+        n_samples: int,
+        discriminator: str = "line",
+        one_hot: bool = True,
+        gradient: float = 1.0,
+        y_intercept: float = 0.0,
+        radius: float = 0.25,
+        seed: int = None,
+    ):
+        """
+        Instantiate the class.
+
+        Parameters
+        ----------
+        n_samples : int
+                Number of samples to generate per class.
+        discriminator : str
+                String to define the discriminator to use.
+                Options are "line" and "circle".
+        one_hot : bool
+                Whether to use one-hot encoding for the classes.
+        gradient : float
+                Gradient of the line, default 1.0.
+        y_intercept : float
+                Intercept of the line, default 1.0.
+        radius : float
+                Radius of the circle, default 0.25.
+        seed : int
+                Random seed for the RNG. Uses a random int if not specified.
+        """
+        self.rng = PRNGKey(seed=seed)
+        self.one_hot = one_hot
+
+        if discriminator == "line":
+            self.discriminator = linear_boundary
+            self.args = (gradient, y_intercept)
+        elif discriminator == "circle":
+            self.discriminator = circle
+            self.args = (radius,)
+        else:
+            raise ValueError("Discriminator not recognised.")
+
+        self.train_ds = self._build_dataset(n_samples=n_samples)
+        self.test_ds = self._build_dataset(n_samples=n_samples)
+
+    def _build_dataset(self, n_samples: int):
+        """
+        Helper method to create datasets quickly.
+
+        Parameters
+        ----------
+        n_samples : int
+                Number of samples to generate per class.
+        """
+        # Create the data-sets
+        data = onp.array(
+            jax.random.uniform(self.rng(), minval=0.0, maxval=1.0, shape=(n_samples, 2))
+        )
+        data = onp.clip(data, 0.0, 1.0)
+        targets = self.discriminator(data, *self.args)  # build classes (0, 1)
+
+        class_one_indices = np.where(targets == 0)[0]
```

Comment:
Is there a reason to work with jax.numpy in the class and with onp in the functions above?

Reply:
What is the problem to identify it?

Reply:
What do you mean